std/src/main/scala/algebra/std/map.scala

```diff
@@ -30,23 +32,98 @@ class MapAdditiveMonoid[K, V](implicit V: AdditiveSemigroup[V]) extends Additive
 
   def plus(xs: Map[K, V], ys: Map[K, V]): Map[K, V] =
     if (xs.size <= ys.size) {
-      addMap(xs, ys)((x, y) => V.plus(x, y))
+      xs.foldLeft(ys) { case (my, (k, x)) =>
+        my.updated(k, my.get(k) match {
+          case Some(y) => V.plus(x, y)
+          case None => x
+        })
+      }
     } else {
-      addMap(ys, xs)((y, x) => V.plus(x, y))
+      ys.foldLeft(xs) { case (mx, (k, y)) =>
+        mx.updated(k, mx.get(k) match {
+          case Some(x) => V.plus(x, y)
+          case None => y
+        })
+      }
+    }
+
+  override def sumN(a: Map[K, V], n: Int): Map[K, V] =
+    if (n > 0) a.map { case (k, v) => (k, V.sumN(v, n)) }
+    else if (n == 0) zero
+    else throw new IllegalArgumentException("Illegal negative exponent to sumN: %s" format n)
+
+  override def sum(as: TraversableOnce[Map[K, V]]): Map[K, V] = {
+    val acc = mutable.Map.empty[K, V]
+    as.foreach { m =>
+      val it = m.iterator
+      while (it.hasNext) {
+        val (k, y) = it.next
+        acc.get(k) match {
+          case None => acc(k) = y
+          case Some(x) => acc(k) = V.plus(x, y)
+        }
+      }
     }
+    StaticMethods.wrapMutableMap(acc)
+  }
 }
 
 class MapSemiring[K, V](implicit V: Semiring[V]) extends MapAdditiveMonoid[K, V] with Semiring[Map[K, V]] {
 
   def times(xs: Map[K, V], ys: Map[K, V]): Map[K, V] =
+    // we figure out which of our maps is smaller, iterate over its
+    // keys, see which of those are in the larger map, and add the
+    // resulting product to our result map.
     if (xs.size <= ys.size) {
-      timesMap(xs, ys)((x, y) => V.times(x, y))
+      xs.foldLeft(Map.empty[K, V]) { case (m, (k, x)) =>
```

Comment:
I wonder if `Map.empty` makes sense here or if we should use the mutable approach? Like, we could allocate an empty `MMap` about the size of the smallest entry (an upperbound).

I *think* this will be faster, if we care, but we can come back to it later if you prefer.